docs: add Mount a Tigris bucket section (#4780)
daytonaio/daytona: apps/docs/src/content/docs/en/mount-external-storage.mdx

</TabItem>
</Tabs>

## Mount a Tigris bucket

> Tigris is S3-compatible, so the same `mount-s3` tool works. Pass `--endpoint-url https://t3.storage.dev` — Tigris uses a single global endpoint with no per-account subdomain. Tigris also supports bucket snapshots and copy-on-write forks via custom request headers, which can be combined with Daytona sandboxes to give each sandbox a writable, isolated copy of a shared bucket without duplicating data.

Reviewer: ```suggestion
Mount a Tigris bucket with the same `mount-s3` tool used for S3. Pass `--endpoint-url https://t3.storage.dev`, because Tigris uses one global endpoint with no per-account subdomain. Tigris also supports bucket snapshots and copy-on-write forks through request headers, so each sandbox can use an isolated writable bucket without duplicating source data.
```

**Credentials** — set `TIGRIS_STORAGE_ACCESS_KEY_ID` and `TIGRIS_STORAGE_SECRET_ACCESS_KEY` in your local environment. The snippets below pass these into the sandbox via `envVars` under the `AWS_ACCESS_KEY_ID` / `AWS_SECRET_ACCESS_KEY` names that `mount-s3` expects.

### Pre-built snapshot

> Build a snapshot with `mount-s3` preinstalled, then launch all Tigris-enabled sandboxes from that snapshot. The mount flow stays identical to S3 except for the Tigris `--endpoint-url`, and startup remains fast because installation is done once at snapshot build time.

Reviewer: ```suggestion
Build a snapshot with `mount-s3` preinstalled, then launch Tigris sandboxes from that snapshot. The mount flow matches S3 except for the Tigris `--endpoint-url`, and startup stays fast because installation happens once during snapshot build.
```

#### Build a snapshot

> Create a reusable snapshot that installs the same `mount-s3` tool used for S3. Tigris is S3-compatible, so this snapshot is identical to S3 setup and only the runtime mount command changes.

Reviewer: ```suggestion
Create a reusable snapshot that installs `mount-s3`. Because Tigris is S3-compatible, this setup matches S3 and only the runtime mount command changes.
```

	Name:  "fuse-tigris",
	Image: image,
})
if err != nil {
	log.Fatal(err)
}
for line := range logChan {
	fmt.Print(line)
}
```

</TabItem>
<TabItem label="Java" icon="seti:java">

```java
import io.daytona.sdk.Daytona;
import io.daytona.sdk.Image;

public class App {
    public static void main(String[] args) {
        try (Daytona daytona = new Daytona()) {
            Image image = Image.base("daytonaio/sandbox")
                .runCommands(
                    "sudo apt-get update "
                        + "&& sudo apt-get install -y --no-install-recommends libfuse2 ca-certificates wget",
                    "arch=\"$(dpkg --print-architecture | sed s/amd64/x86_64/)\" "
                        + "&& wget -O /tmp/mount-s3.deb "
                        + "\"https://s3.amazonaws.com/mountpoint-s3-release/latest/${arch}/mount-s3.deb\" "
                        + "&& sudo apt-get install -y /tmp/mount-s3.deb "
                        + "&& rm /tmp/mount-s3.deb"
                );

            daytona.snapshot().create("fuse-tigris", image, System.out::println);
        }
    }
}
```

</TabItem>
</Tabs>

#### Launch and mount

> Pass your Tigris credentials into the sandbox as `AWS_*` environment variables and mount with the Tigris endpoint URL. This keeps the authentication flow compatible with `mount-s3` while targeting your Tigris account.

Reviewer: ```suggestion
Pass Tigris credentials into the sandbox as `AWS_*` environment variables, then mount with the Tigris endpoint URL. This keeps authentication compatible with `mount-s3` while targeting your Tigris account.
```

import io.daytona.sdk.model.CreateSandboxFromSnapshotParams;
import io.daytona.sdk.model.ExecuteResponse;

import java.util.Map;

public class App {
    public static void main(String[] args) {
        try (Daytona daytona = new Daytona()) {
            CreateSandboxFromSnapshotParams params = new CreateSandboxFromSnapshotParams();
            params.setSnapshot("fuse-tigris");
            params.setEnvVars(Map.of(
                "AWS_ACCESS_KEY_ID", System.getenv("TIGRIS_STORAGE_ACCESS_KEY_ID"),
                "AWS_SECRET_ACCESS_KEY", System.getenv("TIGRIS_STORAGE_SECRET_ACCESS_KEY")
            ));
            Sandbox sandbox = daytona.create(params);

            String mountPath = "/home/daytona/tigris";

            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand(
                "mount-s3 --endpoint-url https://t3.storage.dev "
                    + "my-tigris-bucket " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

### Runtime install

> Start from a default sandbox and install `mount-s3` during startup, then mount your bucket with the Tigris `--endpoint-url`. This path is convenient for prototyping or one-off tasks, but each new sandbox pays the package installation cost.

Reviewer: ```suggestion
Start from a default sandbox, install `mount-s3` during startup, then mount with the Tigris `--endpoint-url`. This path is convenient for prototyping or one-off tasks, but each new sandbox repeats package installation.
```

            Sandbox sandbox = daytona.create(params);

            // Install mount-s3
            sandbox.getProcess().executeCommand(
                "sudo apt-get update "
                    + "&& sudo apt-get install -y --no-install-recommends libfuse2 ca-certificates wget");
            sandbox.getProcess().executeCommand(
                "arch=\"$(dpkg --print-architecture | sed s/amd64/x86_64/)\" "
                    + "&& wget -O /tmp/mount-s3.deb "
                    + "\"https://s3.amazonaws.com/mountpoint-s3-release/latest/${arch}/mount-s3.deb\" "
                    + "&& sudo apt-get install -y /tmp/mount-s3.deb");

            // Mount with Tigris endpoint
            String mountPath = "/home/daytona/tigris";
            sandbox.getProcess().executeCommand(
                "mkdir -p " + mountPath + " && "
                    + "mount-s3 --endpoint-url https://t3.storage.dev "
                    + "my-tigris-bucket " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

### Mount a copy-on-write fork per sandbox

> A Tigris bucket fork is a new bucket created from a snapshot of a source bucket. The fork shares underlying storage with the source until written to — new writes go only to the fork, and the source bucket and other forks are unaffected. Fork creation is constant-time regardless of source bucket size.

Reviewer: ```suggestion
A Tigris bucket fork is a new bucket created from a snapshot of a source bucket. The fork shares underlying storage with the source until it is written to. New writes go only to the fork, and the source bucket and other forks remain unchanged. Fork creation is constant-time regardless of source bucket size.
```

This pattern fits Daytona sandboxes when each sandbox needs a writable copy of a shared dataset (model weights, fixtures, golden data) without duplicating it on every launch.

**Prerequisite** — the source bucket must be created with snapshots enabled. This is a one-time setup, done outside the per-sandbox flow:

```typescript
import { createBucket } from '@tigrisdata/storage'

await createBucket('my-source-bucket', { enableSnapshot: true })
```

> Equivalent in any S3 SDK: send a `CreateBucket` request with the header `X-Tigris-Enable-Snapshot: true`.

Reviewer: ```suggestion
In any S3 SDK, send a `CreateBucket` request with the header `X-Tigris-Enable-Snapshot: true`.
```

> The `@tigrisdata/agent-kit` package wraps the workflow as `createForks()`, which snapshots the source bucket and creates one or more forks in a single call. Passing `credentials: { role: 'Editor' }` also mints a scoped access key per fork — the sandbox gets credentials that can only read/write its own fork bucket, not your whole account.

Reviewer: ```suggestion
The `@tigrisdata/agent-kit` package wraps this workflow as `createForks()`. It snapshots the source bucket and creates one or more forks in a single call. Passing `credentials: { role: 'Editor' }` also creates a scoped access key per fork, so each sandbox can read and write only its own fork bucket instead of the full account.
```

```typescript
import { createForks, teardownForks } from '@tigrisdata/agent-kit'
import { Daytona } from '@daytona/sdk'

const SOURCE_BUCKET = 'my-source-bucket'

// 1. Snapshot the source and create a fork with a scoped access key
const { data: forkSet, error } = await createForks(SOURCE_BUCKET, 1, {
  credentials: { role: 'Editor' },
})
if (error) throw error
const fork = forkSet.forks[0]

// 2. Launch the sandbox with the fork's scoped credentials
const daytona = new Daytona()
const sandbox = await daytona.create({
  snapshot: 'fuse-tigris',
  envVars: {
    AWS_ACCESS_KEY_ID: fork.credentials!.accessKeyId,
    AWS_SECRET_ACCESS_KEY: fork.credentials!.secretAccessKey,
  },
})

// 3. Mount the fork bucket
const mountPath = '/home/daytona/tigris'
await sandbox.process.executeCommand(`mkdir -p ${mountPath}`)
await sandbox.process.executeCommand(
  `mount-s3 --endpoint-url https://t3.storage.dev ${fork.bucket} ${mountPath}`,
)

try {
  const response = await sandbox.process.executeCommand(`ls ${mountPath}`)
  console.log(response.result)
} finally {
  await daytona.delete(sandbox)
  await teardownForks(forkSet) // revokes the scoped key and deletes the fork bucket
}
```

> To drive the same workflow from a language without an `agent-kit` equivalent, use any S3 SDK and send the headers documented below: `X-Tigris-Snapshot: true` on a `CreateBucket` against the source name (capture `X-Tigris-Snapshot-Version` from the response), then `CreateBucket` against the fork name with `X-Tigris-Fork-Source-Bucket` and `X-Tigris-Fork-Source-Bucket-Snapshot` set. Mount the fork using the same `mount-s3` snippets shown above, substituting the fork bucket name.

Reviewer: ```suggestion
To run the same workflow from a language without an `agent-kit` equivalent, use any S3 SDK and send the headers documented below. Send `X-Tigris-Snapshot: true` on `CreateBucket` for the source name, then capture `X-Tigris-Snapshot-Version` from the response. Next, send `CreateBucket` for the fork name with `X-Tigris-Fork-Source-Bucket` and `X-Tigris-Fork-Source-Bucket-Snapshot`. Mount the fork with the same `mount-s3` snippets shown above, replacing the bucket name with the fork bucket.
```

#### Reference: Tigris-specific headers

These headers drive snapshot and fork operations over the S3 API. Use them with any AWS SDK (boto3, aws-sdk-go-v2, aws-sdk-java-v2, aws-sdk-ruby) by attaching a request interceptor.

| Header | Sent on | Purpose |
|---|---|---|
| `X-Tigris-Enable-Snapshot: true` | `CreateBucket` (new source) | Enable snapshots on a new bucket. Required before snapshotting it. |
| `X-Tigris-Snapshot: true` | `CreateBucket` (existing source name) | Take a snapshot of the bucket. Optional `; name=<label>` suffix labels it. |
| `X-Tigris-Snapshot-Version` | Response to snapshot create | Snapshot version ID returned to the caller — pass to fork creation. |
| `X-Tigris-Fork-Source-Bucket` | `CreateBucket` (new fork name) | Source bucket to fork from. |

> | `X-Tigris-Fork-Source-Bucket-Snapshot` | `CreateBucket` (new fork name) | Source snapshot version to fork from. |

Reviewer: ```suggestion
| Header                                     | Sent on                                   | Purpose                                                                        |
| ------------------------------------------ | ----------------------------------------- | ------------------------------------------------------------------------------ |
| **`X-Tigris-Enable-Snapshot: true`**       | **`CreateBucket`** (new source)           | Enable snapshots on a new bucket. Required before snapshotting it.             |
| **`X-Tigris-Snapshot: true`**              | **`CreateBucket`** (existing source name) | Take a snapshot of the bucket. Optional **`; name=<label>`** suffix labels it. |
| **`X-Tigris-Snapshot-Version`**            | Response to snapshot create               | Snapshot version ID returned to the caller — pass to fork creation.            |
| **`X-Tigris-Fork-Source-Bucket`**          | **`CreateBucket`** (new fork name)        | Source bucket to fork from.                                                    |
| **`X-Tigris-Fork-Source-Bucket-Snapshot`** | **`CreateBucket`** (new fork name)        | Source snapshot version to fork from.                                          |
```

> Source: [`tigrisdata/storage`](https://github.com/tigrisdata/storage) — see `shared/headers.ts` for the full header set.

Reviewer: ```suggestion
Source: [**`tigrisdata/storage`**](https://github.com/tigrisdata/storage). See **`shared/headers.ts`** for the full header set.
```

const sandbox = await daytona.create({
  snapshot: 'fuse-tigris',
  envVars: {
    AWS_ACCESS_KEY_ID: fork.credentials!.accessKeyId,
    AWS_SECRET_ACCESS_KEY: fork.credentials!.secretAccessKey,
  },
})

// 3. Mount the fork bucket
const mountPath = '/home/daytona/tigris'
await sandbox.process.executeCommand(`mkdir -p ${mountPath}`)
await sandbox.process.executeCommand(
  `mount-s3 --endpoint-url https://t3.storage.dev ${fork.bucket} ${mountPath}`,
)

try {
  const response = await sandbox.process.executeCommand(`ls ${mountPath}`)
  console.log(response.result)
} finally {
  await daytona.delete(sandbox)
  await teardownForks(forkSet) // revokes the scoped key and deletes the fork bucket
}
```

To run the same workflow from a language without an `agent-kit` equivalent, use any S3 SDK and send the headers documented below. Send `X-Tigris-Snapshot: true` on `CreateBucket` for the source name, then capture `X-Tigris-Snapshot-Version` from the response. Next, send `CreateBucket` for the fork name with `X-Tigris-Fork-Source-Bucket` and `X-Tigris-Fork-Source-Bucket-Snapshot`. Mount the fork with the same `mount-s3` snippets shown above, replacing the bucket name with the fork bucket.

> #### Reference: Tigris-specific headers

Reviewer 2: @davidmyriel Not sure why this section contains 2 tables with duplicated rows and different formatting. Also, the second one has `X-Tigris-Fork-Source-Bucket-Snapshot` header which isn't present in the first one. 

I assume we should only keep the second one here?

Author: Apologies - yes. I just made the change. Hvala :)

Reviewer 2: @davidmyriel Nice, please fix the DCO failing check and we are ready to merge

Author: done
